crux-mir/src/Mir/Concurrency.hs

```diff
@@ -0,0 +1,157 @@
+           pure $! ThreadYield (Acquire ref) refs False
+         _ -> error "TODO: Muxed mutex lock"
+  | matchGeneric "core::crucible::concurrency::mutex_unlock" nm
+  , Ctx.Empty Ctx.:> MirReferenceRepr t <- ctx
+  = do arg <- retrieveTypedArg ctx cf (MirReferenceRepr t) 0
+       let refs = mirRefName arg
+       case refs of
+         [ref] ->
+           pure $! ThreadYield (Release ref) refs False
+         _ -> error "TODO: Muxed mutex unlock"
+  | otherwise = Nothing
+
+mutexName :: BV.BV 32 -> Text.Text
+mutexName bv = Text.pack ("resource-"++ show bv)
+
+mirSpawn :: C.IsSymInterface sym => ExplorePrimitiveMatcher p sym MIR
+mirSpawn _ nm ctx cf@C.CallFrame { C._frameCFG = cfg } _
+  | matchGeneric "core::crucible::concurrency::spawn_internal" nm =
+  -- We use this as both the spawned thread and the function that returns the
+  -- spawned thread's ID. See the note in our libstd/thread/mod.rs
+    do C.Some ty <- retrieveType ctx 0
+       arg       <- retrieveTypedArg ctx cf ty 0
+       let hdl    = C.cfgHandle cfg
+           tpr    = C.BVRepr (W4.knownNat @32)
+           mkRet sym v = W4.bvLit sym W4.knownRepr (BV.mkBV BV.knownNat (fromIntegral v))
+       let check = ( W4.testEquality ctx (Ctx.Empty Ctx.:> ty)
+                   , W4.testEquality (C.handleArgTypes hdl) ctx
+                   , W4.testEquality (C.handleReturnType hdl) tpr
+                   )
+       case check of
+         (Just Refl, Just Refl, Just Refl) ->
+           return $ ThreadCreate ty tpr tpr hdl arg mkRet
+         _ ->
+           error $ "crucible_spawn_internal: unexpected signature "
+                ++ show (ctx, C.handleArgTypes hdl, C.handleReturnType hdl)
+  | otherwise = Nothing
+
+mirExit :: C.IsSymInterface sym => ExplorePrimitiveMatcher p sym MIR
+mirExit _ nm ctx cf _
+  | matchGeneric "core::crucible::concurrency::spawn::thread_exit" nm =
+    do (C.Some idx) <- Ctx.intIndex 0 (Ctx.size ctx)
+       return $ ThreadFinish $ C.Some (C.regVal' (cf ^. C.frameRegs) (C.Reg idx))
+  | otherwise
+  = Nothing
+
+mirRefName :: C.IsSymInterface sym => MirReferenceMux sym tp -> [Text.Text]
+mirRefName t =
+  case t of
+    MirReferenceMux tree ->
+       let refs = fst <$> viewFancyMuxTree tree
+           extr r =
+             case r of
+               MirReference root p ->
+                 Text.pack (show root ++ mirPathName p)
+               -- N.B.: I don't think we should ever hit this:
+               MirReference_Integer t rv ->
+                 error "Unexpected MirReference_Integer in mirRefName"
+       in (extr <$> refs)
+
+mirPathName :: C.IsSymInterface sym => MirReferencePath sym tpr t -> String
+mirPathName p =
+  case p of
+    Empty_RefPath -> ""
+    Any_RefPath _ p -> mirPathName p
+    Field_RefPath _ p idx -> mirPathName p ++ "." ++ show (indexVal idx)
+    Variant_RefPath _ p idx -> mirPathName p ++ "." ++ show (indexVal idx)
+    Index_RefPath _ p idx -> mirPathName p ++ "[" ++ show (W4.printSymExpr idx) ++ "]"
```

Comment:
Are you going to run into trouble here if you ever get two `idx`s that are syntactically distinct but semantically equal?  Imagine having symbolic integers `i` and `j` where the path condition implies that `i == j`.  `&arr[i]` and `&arr[j]` point to the same thing, but will produce different strings here.

Reply:
Yes! That's an issue -- the right resolution for now is to simply treat all elements as potentially interfering.